File: binary_tree_maximum_path_sum/BinaryTreeMaximumPathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class BinaryTreeMaximumPathSum {
    public int maxPathSum(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        int[] maxSum = new int[]{Integer.MIN_VALUE};
        dfs(root, maxSum);
        return maxSum[0];
    }
    
    private int dfs(TreeNode root, int[] maxSum) {
        if (root == null) {
            return 0;
        }
        
        int left = Math.max(0, dfs(root.left, maxSum));
        int right = Math.max(0, dfs(root.right, maxSum));
        maxSum[0] = Math.max(maxSum[0], left + right + root.val);
        return Math.max(left, right) + root.val;
    }
}

File: binary_tree_maximum_path_sum/Binary_Tree_Maximum_Path_Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Binary_Tree_Maximum_Path_Sum {
public:
    int maxPathSum(TreeNode* root) {
        if (root == NULL) {
            return 0;
        }
        
        int max_sum = INT_MIN;
        dfs(root, max_sum);
        return max_sum;
    }
    
private:
    int dfs(TreeNode* root, int& max_sum) {
        if (root == NULL) {
            return 0;
        }
        
        int left = max(0, dfs(root->left, max_sum));
        int right = max(0, dfs(root->right, max_sum));
        max_sum = max(max_sum, root->val + left + right);
        return root->val + max(left, right);
    }
};
